measure_definitions.py

```diff
@@ -0,0 +1,165 @@
+# the accusation that this is all a bit Java and it would be more readable if they were
+# lowercase.
+
+
+# DEMOGRAPHICS
+#
+
+practice = practice_registrations.for_patient_on(INTERVAL_START_DATE)
+
+has_died = ons_deaths.where(ons_deaths.date <= INTERVAL_START_DATE).exists_for_patient()
+
+age = patients.age_on(INTERVAL_START_DATE)
+
+age_band = case(
+    when((age >= 0) & (age <= 4)).then("0-4"),
+    when((age >= 5) & (age <= 9)).then("5-9"),
+    when((age >= 10) & (age <= 14)).then("10-14"),
+    when((age >= 15) & (age <= 19)).then("15-19"),
+    when((age >= 20) & (age <= 24)).then("20-24"),
+    when((age >= 25) & (age <= 29)).then("25-29"),
+    when((age >= 30) & (age <= 34)).then("30-34"),
+    when((age >= 35) & (age <= 39)).then("35-39"),
+    when((age >= 40) & (age <= 44)).then("40-44"),
+    when((age >= 45) & (age <= 49)).then("45-49"),
+    when((age >= 50) & (age <= 54)).then("50-54"),
+    when((age >= 55) & (age <= 59)).then("55-59"),
+    when((age >= 60) & (age <= 64)).then("60-64"),
+    when((age >= 65) & (age <= 69)).then("65-69"),
+    when((age >= 70) & (age <= 74)).then("70-74"),
+    when((age >= 75) & (age <= 79)).then("75-79"),
+    when((age >= 80) & (age <= 84)).then("80-84"),
+    when((age >= 85) & (age <= 89)).then("85-89"),
+    when(age >= 90).then("90plus"),
+    default="missing",
+)
+
+
+# DENOMINATOR
+#
+
+population = (
+    practice.exists_for_patient()
+    & ~has_died
+    & patients.sex.is_in(["male", "female"])
+    & (age_band != "missing")
+)
+
+
+# NUMERATOR
+#
+# Patients who have had a matching event in the interval
+
+# https://www.opencodelists.org/codelist/primis-covid19-vacc-uptake/cov1decl/v1.1/
+codelist = [
+    "1324721000000108",  # Severe acute respiratory syndrome coronavirus 2 vaccination dose declined
+    "1324741000000101",  # Severe acute respiratory syndrome coronavirus 2 vaccination first dose declined
+    "1324811000000107",  # Severe acute respiratory syndrome coronavirus 2 immunisation course declined
+]
+
+e = clinical_events
+event = (
+    e.where(e.snomedct_code.is_in(codelist))
+    .where(e.date.is_on_or_between(INTERVAL_START_DATE, INTERVAL_END_DATE))
+    .sort_by(e.date)
+    .first_for_patient()
+)
+had_matching_event = event.exists_for_patient()
+
+
+# DATE RANGE
+#
+# The `Intervals` methods are helper functions which return a list of date pairs, being
+# the start and ends (inclusive) of the specified intervals e.g.
+#
+#     Intervals.weeks_between("2020-01-01", "2020-01-15") == [
+#         (date(2020,1,1), date(2020,1,7)),
+#         (date(2020,1,8), date(2020,1,14)),
+#         (date(2020,1,15), date(2020,1,21)),
+#     ]
+#
+# `Intervals` is just acting as a namespace here; it felt cleaner to me to keep these
+# together rather than having them as free floating functions. But I'm open to other
+# suggestions.
+
+intervals = Intervals.months_between("2020-12-01", "2021-04-01")
```

Comment:
I worry slightly about inclusive/exclusive ambiguity. Is there a way to resolve that? I can't think of any!